tests/_models/test_package.py

```diff
@@ -19,34 +22,66 @@ def sort_map_values(m: dict[str, Any]) -> dict[str, Any]:
 
 def test_guess_version_setuptools(mocker: MockerFixture) -> None:
     mocker.patch("pipdeptree._models.package.version", side_effect=PackageNotFoundError)
-    result = ReqPackage(mocker.MagicMock(key="setuptools")).installed_version
+    r = MagicMock()
+    r.name = "setuptools"
+    result = ReqPackage(r).installed_version
     assert result == "?"
 
 
 def test_dist_package_render_as_root() -> None:
-    foo = Mock(key="foo", project_name="foo", version="20.4.1")
+    foo = Mock(metadata={"Name": "foo"}, version="20.4.1")
     dp = DistPackage(foo)
     is_frozen = False
     assert dp.render_as_root(frozen=is_frozen) == "foo==20.4.1"
 
 
 def test_dist_package_render_as_branch() -> None:
-    foo = Mock(key="foo", project_name="foo", version="20.4.1")
-    bar = Mock(key="bar", project_name="bar", version="4.1.0")
-    bar_req = Mock(key="bar", project_name="bar", version="4.1.0", specs=[(">=", "4.0")])
+    foo = Mock(metadata={"Name": "foo"}, version="20.4.1")
+    bar = Mock(metadata={"Name": "bar"}, version="4.1.0")
+    bar_req = MagicMock(version="4.1.0", specifier=[">=4.0"])
+    bar_req.name = "bar"
     rp = ReqPackage(bar_req, dist=bar)
     dp = DistPackage(foo).as_parent_of(rp)
     is_frozen = False
     assert dp.render_as_branch(frozen=is_frozen) == "foo==20.4.1 [requires: bar>=4.0]"
 
 
+def test_dist_package_render_as_root_with_frozen() -> None:
+    json_text = '{"dir_info": {"editable": true}, "url": "file:///A/B/foo"}'
+    foo = Mock(metadata={"Name": "foo"}, version="20.4.1")
+    foo.read_text = Mock(return_value=json_text)
+    dp = DistPackage(foo)
+    is_frozen = True
+    expect = "# Editable install with no version control (foo===20.4.1)\n-e /A/B/foo"
+    assert dp.render_as_root(frozen=is_frozen) == expect
+
+
+def test_dist_package_render_as_root_with_frozen_none_url() -> None:
+    temp_dir = tempfile.TemporaryDirectory()
+    file_path = Path(temp_dir.name) / "foo.egg-link"
+    with Path(file_path).open("w") as f:
+        f.write("/A/B/foo")
+
+    sys.path.append(temp_dir.name)
+    json_text = '{"dir_info": {"editable": true}'
+    foo = Mock(metadata={"Name": "foo"}, version="20.4.1")
+    foo.read_text = Mock(return_value=json_text)
+    dp = DistPackage(foo)
+    is_frozen = True
+    expect = "# Editable install with no version control (foo===20.4.1)\n-e /A/B/foo"
+    assert dp.render_as_root(frozen=is_frozen) == expect
+    sys.path.remove(temp_dir.name)
+    temp_dir.cleanup()
```

Comment:
I think it would be nice to have tests for frozen requirements to make sure that our code and pip keeps behaving nicely together. Though I'm not quite sure how that would best look right now. Anyways, a couple things:
- Test `Package.as_frozen_repr()` directly instead of having to call it through `Package.render_as_root()`
- Use a pytest [fixture](https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmp-path-fixture) for creating a temporary directory
- Use `pytest.MonkeyPatch`to change `sys.path` as it's safer
- It looks like this particular test is trying to test the egg-link path in `FrozenRequirement.from_dist()` but you have included 
- (Nitpick) Use `expected` instead of `expect`

Reply:
> Test Package.as_frozen_repr() directly instead of having to call it through DistPackage.render_as_root() or ReqPackage.render_as_root()

I tried to call `Package.as_frozen_repr()` directly, but got the following error.
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python3.11/site-packages/pipdeptree/_models/package.py:94: in as_frozen_repr
    from pip._internal.operations.freeze import FrozenRequirement  # noqa: PLC0415, PLC2701 # pragma: no cover
/usr/lib/python3.11/site-packages/pip/_internal/operations/freeze.py:11: in <module>
    from pip._internal.req.constructors import (
/usr/lib/python3.11/site-packages/pip/_internal/req/__init__.py:7: in <module>
    from .req_file import parse_requirements
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    """
    Requirements file parsing
    """
    
    import logging
>   import optparse
E   ModuleNotFoundError: No module named 'optparse'

/usr/lib/python3.11/site-packages/pip/_internal/req/req_file.py:6: ModuleNotFoundError
```

Reply:
> It looks like this particular test is trying to test the egg-link path in FrozenRequirement.from_dist() but you have included

`test_dist_package_render_as_root_with_frozen` will not run into `.egg-link` branch since `self.direct_url` exists. 
This test is for the situations like `python setup.py develop`.

Reply:
> I tried to call Package.as_frozen_repr() directly, but got the following error.

Not sure why you would have this problem because these tests still eventually call this and also, weirdly enough, optparse is apart of the [standard library](https://github.com/python/cpython/blob/3.11/Lib/optparse.py) so I'm not sure why you get a `ModuleNotFoundError`.

Reply:
> I tried to call Package.as_frozen_repr() directly, but got the following error.

Figured it out.
We can not use`monkeypatch.setattr` for `sys.path` directly. Instead, we use `syspath_prepend` see [here](https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.syspath_prepend)

Reply:
Another problem is that is we use `Package.as_frozen_repr()`, `return self.render_as_root(frozen=frozen)` this line won't be covered.
So I suggest keeping using `render_as_root` API..

Reply:
Something else we could do:
- Add the tests for `Package.as_frozen_repr()` that includes the code paths you were trying to cover in `FrozenRequirement.from_dist()`
- Keep your existing `render_as_root()` tests, but since we will have tests for `Package.as_frozen_repr()`, we can mock its return value. The benefit of this is that we cover all of these functions with less test code duplication

Reply:
It may be not necessary to add a test only for `return self.render_as_root(frozen=frozen)`.
I made this change to get rid of the coverage problem.
```
     def render_as_root(self, *, frozen: bool) -> str:
-        if not frozen:
-            return f"{self.project_name}=={self.version}"
-        return self.as_frozen_repr(self)
+        return self.as_frozen_repr(self) if frozen else f"{self.project_name}=={self.version}"
```
Let's do what you said here.

> Test Package.as_frozen_repr() directly instead of having to call it through DistPackage.render_as_root() or ReqPackage.render_as_root()